### PR TITLE
TEP-0142: Add support for params between Step and StepActions

### PR DESCRIPTION
**Pull request by chitrangpatel**
Following the previous [PR](https://github.com/tektoncd/pipeline/pull/7334), which introduced the syntax in `Steps` to provide params to `StepAction`, this PR reconciles `param` usage between `Steps` and `StepActions`. This completes support for params in `StepActions`. This work is part of issue https://github.com/tektoncd/pipeline/issues/7259.

# Changes

We how pass objectParam replacements inside a `Task` since the `Steps` can provide params to `StepActions`. The params passed to `StepActions` could be object params and could require objectParam replacements.

For example, this is now allowed:
```
steps:
  - ref:
      name: step-action
  params:
    - name: object-param
       value: $(params.objectparam[*])
```
Therefore, we now need to allow passing of objectParams into a `Task`. 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
[WIP] Completes support for params in StepActions.
```

/kind feature

### CODE REVIEW

File: pkg/reconciler/pipelinerun/resources/apply.go [hunk 1]
```diff
@@ -367,9 +367,9 @@ func propagateParams(t v1.PipelineTask, stringReplacements map[string]string, ar
 				}
 			}
 		}
-		t.TaskSpec.TaskSpec = *resources.ApplyReplacements(&t.TaskSpec.TaskSpec, stringReplacementsDup, arrayReplacementsDup)
+		t.TaskSpec.TaskSpec = *resources.ApplyReplacements(&t.TaskSpec.TaskSpec, stringReplacementsDup, arrayReplacementsDup, objectReplacementsDup)
```
Comment (jerop) [inline pkg/reconciler/pipelinerun/resources/apply.go:370]: is this implementing propagation of object parameters? if so, i think this should be in a separate PR

Comment (chitrangpatel) [reply pkg/reconciler/pipelinerun/resources/apply.go:370]: Yes and no. 

For an inlined-step this is not needed at all since inside the Task, we cannot really pass complete objects. Users need to declare explicit keys of the object they want to use. 

e.g. This is allowed
```
script: |
  echo $(params.objectparam.key)
```

but this is not:
```
script: |
  echo $(params.objectparam[*])
```
For this reason, we never passed object params for replacements inside a Task.

However now, with `StepActions`, a `Step` will need to pass it object param completely. 
For example, this is allowed:
```
steps:
  - ref:
      name: step-action
  params:
    - name: object-param
       value: $(params.objectparam[*])
```
Therefore, we now need to allow passing of objectParams into a Task. 
If you look at the [usage](https://github.com/tektoncd/pipeline/pull/7332/files#diff-487eb42be2c62df773ee4e76e5570d045aeac1ac0c7d56e095ebf14a7deac58bR374-R376), we only limit it to the use for `params` in `Steps`, and nowhere else so that the existing behaviour for inlined step remains intact.

File: pkg/reconciler/taskrun/resources/apply.go [hunk 2]
```diff
@@ -67,6 +127,9 @@ func ApplyParameters(ctx context.Context, spec *v1.TaskSpec, tr *v1.TaskRun, def
 					arrayReplacements[fmt.Sprintf(pattern, p.Name)] = p.Default.ArrayVal
 				}
 			case v1.ParamTypeObject:
+				for _, pattern := range paramPatterns {
+					objectReplacements[fmt.Sprintf(pattern, p.Name)] = p.Default.ObjectVal
+				}
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go:132]: Yes, like I mentioned in my previous comment, it is needed now inside a Task. For inlined Steps, it was not needed before. Therefore, I don't think we can split this into another PR since it would be meaningless.

File: pkg/reconciler/pipelinerun/resources/apply.go (hunk 1 above)
Comment (jerop) [reply pkg/reconciler/pipelinerun/resources/apply.go:370]: Makes sense, thanks for the detailed explanation. Please add this note in the commit message so that we have a record of why this change is included 🙏🏾

File: pkg/reconciler/taskrun/resources/apply.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jerop) [reply pkg/reconciler/taskrun/resources/apply.go:132]: Sounds good 👍🏾

File: pkg/reconciler/pipelinerun/resources/apply.go (hunk 1 above)
Comment (chitrangpatel) [reply pkg/reconciler/pipelinerun/resources/apply.go:370]: Done! I added it to changes section.

File: pkg/reconciler/taskrun/resources/apply.go [hunk 3]
```diff
@@ -46,14 +46,73 @@ var (
...
-func ApplyParameters(ctx context.Context, spec *v1.TaskSpec, tr *v1.TaskRun, defaults ...v1.ParamSpec) *v1.TaskSpec {
+// applyStepActionParameters applies the params from a TaskRun.Input.Parameters to a TaskSpec
+func applyStepActionParameters(ctx context.Context, step *v1.Step, spec *v1.TaskSpec, tr *v1.TaskRun, stepParams v1.Params, defaults []v1.ParamSpec) *v1.Step {
+	if stepParams != nil {
+		stringR, arrayR, objectR := getTaskParameters(ctx, spec, tr, spec.Params...)
+		stepParams = stepParams.ReplaceVariables(stringR, arrayR, objectR)
+	}
+	stringReplacements := map[string]string{}
+	arrayReplacements := map[string][]string{}
+	objectReplacements := map[string]map[string]string{}
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/apply.go:57]: is objectReplacements used in this function? Looks like it is created, updated but not used in container.ApplyStepReplacements

File: examples/v1/pipelineruns/alpha/stepaction-params.yaml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        key2: "step-action default key2"
+        key3: "step-action default key3"
+  image: bash:3.2
+  args: [
+    "echo",
+    "$(params.array-param[*])",
+    "$(params.string-param)",
+    "$(params.object-param.key1)",
+    "$(params.object-param.key2)"
+  ]
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply examples/v1/pipelineruns/alpha/stepaction-params.yaml:35]: Good idea!

File: pkg/reconciler/taskrun/resources/apply.go [hunk 5]
```diff
@@ -46,14 +46,73 @@ var (
 	}
 )
 
-// ApplyParameters applies the params from a TaskRun.Input.Parameters to a TaskSpec
-func ApplyParameters(ctx context.Context, spec *v1.TaskSpec, tr *v1.TaskRun, defaults ...v1.ParamSpec) *v1.TaskSpec {
+// applyStepActionParameters applies the params from a TaskRun.Input.Parameters to a TaskSpec
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go:49]: ah yes!

File: pkg/reconciler/taskrun/resources/apply.go (hunk 3 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go:57]: Yes, good catch. I cleaned up the usage but forgot this.

File: pkg/reconciler/taskrun/resources/apply.go [hunk 6]
```diff
@@ -46,14 +46,73 @@ var (
...
+func applyStepActionParameters(ctx context.Context, step *v1.Step, spec *v1.TaskSpec, tr *v1.TaskRun, stepParams v1.Params, defaults []v1.ParamSpec) *v1.Step {
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/apply.go:50]: It looks like there is a lot of repetitive code with `ApplyParameters`, I think one followup is to cleanup the code if possible

File: pkg/reconciler/taskrun/resources/apply.go [hunk 7]
```diff
@@ -46,14 +46,73 @@ var (
...
+			default:
+				for _, pattern := range paramPatterns {
+					stringReplacements[fmt.Sprintf(pattern, p.Name)] = p.Default.StringVal
+				}
+			}
+		}
+	}
+
+	// Set and overwrite params with the ones from the Step
+	stepStrings, stepArrays, stepObjects := paramsFromStep(ctx, stepParams)
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/apply.go:88]: I think `paramsFromStep` and current `paramsFromTaskRun` can be merged into one if we don't need the
`stepObjects` here (looks like it is not used because objectReplacements is not used) and change  `paramsFromTaskRun` to accept the v1.Params instead of a `taskrun`

File: pkg/reconciler/taskrun/resources/apply.go (hunk 6 above)
Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/apply.go:50]: Or we can just do it in this PR.

File: examples/v1/pipelineruns/alpha/stepaction-params.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply examples/v1/pipelineruns/alpha/stepaction-params.yaml:35]: fixed

File: pkg/reconciler/taskrun/resources/apply.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go]: fixed

File: pkg/reconciler/taskrun/resources/apply.go (hunk 3 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go]: fixed

File: pkg/reconciler/taskrun/resources/apply.go (hunk 7 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go]: paramsFromTaskRun needs to output objectParams but paramsFromStep does not.

File: pkg/reconciler/taskrun/resources/apply.go (hunk 6 above)
Comment (jerop) [reply pkg/reconciler/taskrun/resources/apply.go:50]: I recommend that any cleanup should happen in a separate PR that is merged before this one 🙏🏾

Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go]: I don't think that would be necessary since the cleanup is quite small and irrelevant without the need for step actions. PTAL. I'm happy to separate it into a different PR if you still think so.

File: pkg/reconciler/taskrun/resources/apply.go (hunk 7 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go]: cleaned up all around. PTAL.

File: pkg/reconciler/taskrun/resources/taskspec_test.go [hunk 8]
```diff
@@ -471,10 +757,10 @@ func TestGetStepActionsData(t *testing.T) {
 
 func TestGetStepActionsData_Error(t *testing.T) {
 	tests := []struct {
-		name           string
-		tr             *v1.TaskRun
-		stepActionFunc func(ctx context.Context, ref *v1.Ref) (*v1alpha1.StepAction, *v1.RefSource, error)
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/taskspec_test.go:476]: Thanks!

File: pkg/reconciler/taskrun/resources/validate_params.go [hunk 9]
```diff
@@ -45,3 +46,40 @@ func ValidateOutOfBoundArrayParams(declarations v1.ParamSpecs, params v1.Params,
 	}
 	return nil
 }
+
+func validateStepHasStepActionParameters(ctx context.Context, stepParams v1.Params, stepActionDefaults []v1.ParamSpec) error {
```
Comment (Yongxuanzhang) [inline pkg/reconciler/taskrun/resources/validate_params.go:50]: looks like the ctx is not used?

Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/validate_params.go:50]: I wonder if we have the same logic from Task? Since we are just passing v1.Params and v1.ParamSpec and they are not specific for Step. I wonder if we could reuse some code. But I cannot find it atm

**Review by Yongxuanzhang — APPROVED** [review]
Thanks! Just some minor comments/questions

File: pkg/reconciler/taskrun/resources/apply.go (hunk 6 above)
Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/apply.go]: I was about to open that PR but saw that the changes have already been made in this one.  😄 If it is a separate we can also merge it very quickly.
https://github.com/Yongxuanzhang/pipeline/tree/refactor-taskrun-apply

File: pkg/reconciler/taskrun/resources/validate_params.go (hunk 9 above)
Comment (Yongxuanzhang) [reply pkg/reconciler/taskrun/resources/validate_params.go:50]: Oh might be this one: https://github.com/tektoncd/pipeline/blob/a106110878496a25e4a863b19cd8f13a7369843c/pkg/reconciler/taskrun/validate_taskrun.go#L36

Do you think we can reuse it?

File: pkg/reconciler/taskrun/resources/apply_test.go [hunk 10]
```diff
@@ -175,6 +175,26 @@ var (
 		}},
 	}
 
+	stepParamTaskSpec = &v1.TaskSpec{
+		Params: v1.ParamSpecs{{
+			Name: "myObject",
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply_test.go:180]: done.

File: pkg/reconciler/taskrun/resources/apply.go [hunk 11]
```diff
@@ -46,14 +46,66 @@ var (
...
+	}
+	for k, v := range stepArrays {
+		arrayReplacements[k] = v
+	}
+
+	container.ApplyStepReplacements(step, stringReplacements, arrayReplacements)
+	return step
+}
+
+// getTaskParameters applies the params from a TaskRun.Input.Parameters to a TaskSpec
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go:71]: done

File: pkg/reconciler/taskrun/resources/apply.go [hunk 12]
```diff
@@ -46,14 +46,66 @@ var (
...
+				} else {
+					objectReplacements[k] = v
+				}
+			}
+		}
+	}
+	return stringReplacements, arrayReplacements, objectReplacements
+}
+
+// ApplyParameters applies the params from a TaskRun.Input.Parameters to a TaskSpec
```
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/apply.go:99]: done

File: pkg/reconciler/taskrun/resources/validate_params.go (hunk 9 above)
Comment (chitrangpatel) [reply pkg/reconciler/taskrun/resources/validate_params.go:50]: Not quite because in the function `validateStepHasStepActionParameters ` we cannot validate the types and properties yet since the parameter replacements have not yet happened. Here, we want to ensure that the Step is passing params to the StepAction. The exactly value and type of the param passed by Step is still unknown since it is likely something like `$(params.foo)` instead of the resolved value.

File: examples/v1/pipelineruns/alpha/stepaction-params.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (chitrangpatel) [reply examples/v1/pipelineruns/alpha/stepaction-params.yaml:35]: done

**Review by jerop — COMMENTED** [review]
🎉 

/lgtm
